core/src/main/java/com/gorisse/thomas/sceneform/environment/KTXEnvironment.kt

```diff
@@ -0,0 +1,105 @@
+package com.gorisse.thomas.sceneform.environment
+
+import android.content.Context
+import androidx.lifecycle.LifecycleCoroutineScope
+import com.google.android.filament.IndirectLight
+import com.google.android.filament.Skybox
+import com.google.android.filament.utils.KTXLoader
+import com.gorisse.thomas.sceneform.filament.Filament
+import com.gorisse.thomas.sceneform.util.fileBuffer
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.nio.Buffer
+
+class KTXEnvironment(
+    indirectLight: IndirectLight?,
+    sphericalHarmonics: FloatArray? = null,
+    skybox: Skybox? = null
+) : Environment(
+    indirectLight = indirectLight,
+    sphericalHarmonics = sphericalHarmonics,
+    skybox = skybox
+)
+
+/**
+ * ### Utility for producing environment resources from precompiled cmgen generated KTX files
+ *
+ * [Documentation][KTXLoader.createEnvironment]
+ *
+ * @param iblKtxFileLocation the ibl file location
+ * [Documentation][com.google.ar.sceneform.util.ResourceLoader.fileBuffer]
+ * @param skyboxKtxFileLocation the skybox file location
+ * [Documentation][com.google.ar.sceneform.util.ResourceLoader.fileBuffer]
+ *
+ * @return [Documentation][KTXLoader.createEnvironment]
+ */
+@JvmOverloads
+suspend fun KTXLoader.loadEnvironment(
+    context: Context,
+    iblKtxFileLocation: String,
+    skyboxKtxFileLocation: String? = null
+): Environment? {
+    var environment: Environment? = null
+    return try {
+        val ibl = context.fileBuffer(iblKtxFileLocation)
+        val skybox = skyboxKtxFileLocation?.let { context.fileBuffer(it) }
+        withContext(Dispatchers.Main) {
+            createEnvironment(ibl, skybox)
+                .also { environment = it }
+        }
```

Comment:
@romainguy Why can't I run it on `Dispatchers.IO`?

Reply:
Because all Filament APIs must be invoked from the same thread (or you need to perform your own synchronization). It doesn't have to be the main thread though, just the same thread.